sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -205,30 +206,184 @@ class Analyzer(
+            resolved :+
+              name -> executeSameContext(substituteCTE(relation, resolved, recursiveTableName))
+        }, None)
       case other => other
     }
 
-    def substituteCTE(plan: LogicalPlan, cteRelations: Seq[(String, LogicalPlan)]): LogicalPlan = {
-      plan resolveOperatorsDown {
-        case u: UnresolvedRelation =>
-          cteRelations.find(x => resolver(x._1, u.tableIdentifier.table))
-            .map(_._2).getOrElse(u)
-        case other =>
-          // This cannot be done in ResolveSubquery because ResolveSubquery does not know the CTE.
-          other transformExpressions {
-            case e: SubqueryExpression =>
-              e.withNewPlan(substituteCTE(e.plan, cteRelations))
+    def substituteCTE(
+        plan: LogicalPlan,
+        cteRelations: Seq[(String, LogicalPlan)],
+        recursiveTableName: Option[String]): LogicalPlan = {
+      def substitute(
+          plan: LogicalPlan,
+          inSubQuery: Boolean = false): (LogicalPlan, Boolean) = {
+        val references = mutable.Set.empty[UnresolvedRecursiveReference]
+
+        def newReference(recursiveTableName: String) = {
+          val recursiveReference = UnresolvedRecursiveReference(recursiveTableName)
+          references += recursiveReference
+
+          recursiveReference
+        }
+
+        val newPlan = plan resolveOperatorsDown {
+          case u: UnresolvedRelation =>
+            val table = u.tableIdentifier.table
+
+            val recursiveReference = recursiveTableName.find(resolver(_, table)).map { name =>
+              if (inSubQuery) {
+                throw new AnalysisException(
+                  s"Recursive reference ${name} can't be used in a subquery")
+              }
+
+              newReference(name)
+            }
+
+            recursiveReference
+              .orElse(cteRelations.find(x => resolver(x._1, table)).map(_._2))
+              .getOrElse(u)
+
+          case other =>
+            // This cannot be done in ResolveSubquery because ResolveSubquery does not know the CTE.
+            other transformExpressions {
+              case e: SubqueryExpression => e.withNewPlan(substitute(e.plan, true)._1)
+            }
+        }
+
+        (newPlan, !references.isEmpty)
+      }
+
+      plan match {
+        case SubqueryAlias(name, u: Union) if recursiveTableName.isDefined =>
+          def combineUnions(union: Union): Seq[LogicalPlan] = union.children.flatMap {
+            case u: Union => combineUnions(u)
+            case o => Seq(o)
           }
+
+          val substitutedTerms = combineUnions(u).map(substitute(_))
+          val (anchorTerms, recursiveTerms) = substitutedTerms.partition(!_._2)
+
+          if (!recursiveTerms.isEmpty) {
+            if (anchorTerms.isEmpty) {
+              throw new AnalysisException(s"There should be at least 1 anchor term defined in a " +
+                s"recursive query $name")
+            }
+
+            case class PlanTraverseStatus(
```

Comment:
instead of creating this status class, can't we use a boolean: `isAllowedRecursiceReference`? We may be less precise when telling the user the reason, but the code would be much simpler and concise.